File: lib/ons/producer.rb
module Ons
  # the ONS Producer
  class Producer
    # Get all the Producer instances.
    #
    # @return [<Producer>] all the Producer instances
    def self.instances
      @instances ||= []
    end

    # Create a new aliyun ONS Producer instance.
    #
    # @param access_key [String] the access key to aliyun ONS
    # @param secret_key [String] the secret key to aliyun ONS
    # @param producer_id [String] the producer ID
    # @param options [Hash{String, Symbol => String}]
    # @option options [String] :namesrv_addr the nameserver used to fetching ons_addr
    # @option options [String] :ons_addr the ONS server address
    # @option options [String, Fixnum, Bignum] :send_timeout (3000) send message timeout
    def initialize(access_key, secret_key, producer_id, options = {})
      @producer = Internal::Producer.new(access_key, secret_key, producer_id, options)

      # register instance
      self.class.instances << self
    end

    # Send a message.
    #
    # @example send 'Hello, World!' message with tag :tagA which is under topic :TopicTestMQ
    #   producer.send_message('TopicTestMQ', 'tagA', 'Hello, World!')
    #
    # @see #send_timer_message
    #
    # @param topic [String] the message topic
    # @param tag [String] the message tag
    # @param body [String] the message body
    # @param key [String] the message key
    # @return [String] the message id
    def send_message(topic, tag, body, key = '')
      @producer.send_message(topic, tag, body, key)
    end

    # Send a timer message.
    #
    # @example send 'Hello, World!' message at 30 seconds later
    #   producer.send_timer_message('TopicTestMQ', 'tagA', 'Hello, World!', Time.now + 30)
    #
    # @see #send_message
    #
    # @param topic [String] the message topic
    # @param tag [String] the message tag
    # @param body [String] the message body
    # @param timer [#to_i] when deliver the message
    # @param key [String] the message key
    # @return [String] the message id
    def send_timer_message(topic, tag, body, timer, key = '')
      @producer.send_timer_message(topic, tag, body, timer.to_i * 1000, key)
    end

    # Start the Producer instance.
    #
    # @return [self] returns itself
    #
    # @note this method should be called before send any message.
    def start
      @producer.start
      self
    end

    # Shutdown the Producer instance
    #
    # @return [void]
    #
    # @note this method should be called before program exit, otherwise it would case a memory leak.
    # Please see also {.register_cleanup_hooks} if you want call it automatically.
    def shutdown
      @producer.shutdown
    end
  end
end

File: ext/ons/producer.cpp
#include <rice/Constructor.hpp>
#include <rice/Data_Type.hpp>
#include <rice/Hash.hpp>
#include <rice/Module.hpp>
#include <rice/Object.hpp>
#include <rice/String.hpp>
#include <rice/Symbol.hpp>
#include <ons/ONSFactory.h>
#include "producer.hpp"

Producer::Producer(Rice::String accessKey, Rice::String secretKey, Rice::String producerId, Rice::Hash options)
{
  ons::ONSFactoryProperty factoryInfo;
  factoryInfo.setFactoryProperty(ons::ONSFactoryProperty::AccessKey,   accessKey.str());
  factoryInfo.setFactoryProperty(ons::ONSFactoryProperty::SecretKey,   secretKey.str());
  factoryInfo.setFactoryProperty(ons::ONSFactoryProperty::ProducerId, producerId.str());

  Rice::Object namesrvAddr = options.call("[]", Rice::String("namesrv_addr"));
  if (!namesrvAddr) { namesrvAddr = options.call("[]", Rice::Symbol("namesrv_addr")); }
  if (namesrvAddr) { factoryInfo.setFactoryProperty(ons::ONSFactoryProperty::NAMESRV_ADDR, ((Rice::String)namesrvAddr).str()); }

  Rice::Object onsAddr = options.call("[]", Rice::String("ons_addr"));
  if (!onsAddr) { onsAddr = options.call("[]", Rice::Symbol("ons_addr")); }
  if (onsAddr) { factoryInfo.setFactoryProperty(ons::ONSFactoryProperty::ONSAddr, ((Rice::String)onsAddr).str()); }

  Rice::Object sendTimeout = options.call("[]", Rice::String("send_timeout"));
  if (!sendTimeout) { sendTimeout = options.call("[]", Rice::Symbol("send_timeout")); }
  if (sendTimeout.rb_type() == T_FIXNUM || sendTimeout.rb_type() == T_BIGNUM) { sendTimeout = sendTimeout.to_s(); }
  if (sendTimeout) { factoryInfo.setFactoryProperty(ons::ONSFactoryProperty::SendMsgTimeoutMillis, ((Rice::String)sendTimeout).str()); }

  this->onsProducer = ons::ONSFactory::getInstance()->createProducer(factoryInfo);
}

Rice::String Producer::sendMessage(Rice::String topic, Rice::String tag, Rice::String body, Rice::String key)
{
  ons::Message message(topic.str(), tag.str(), body.str());
  if (key.length() != 0) { message.setKey(key.str()); }

  ons::SendResultONS sendResult = this->onsProducer->send(message);
  return sendResult.getMessageId();
}

Rice::String Producer::sendTimerMessage(Rice::String topic, Rice::String tag, Rice::String body, Rice::Object deliverTimestamp, Rice::String key)
{
  ons::Message message(topic.str(), tag.str(), body.str());
  if (key.length() != 0) { message.setKey(key.str()); }

  long long timestamp = NUM2LL(deliverTimestamp.value());
  if (timestamp != 0) { message.setStartDeliverTime(timestamp); }

  ons::SendResultONS sendResult = this->onsProducer->send(message);
  return sendResult.getMessageId();
}

void Producer::start()
{
  this->onsProducer->start();
}

void Producer::shutdown()
{
  this->onsProducer->shutdown();
}

void define_class_producer_under_module(Rice::Module module)
{
  Rice::Data_Type<Producer> rb_cProducer = Rice::define_class_under<Producer>(module, "Producer");

  rb_cProducer.define_constructor(Rice::Constructor<Producer, Rice::String, Rice::String, Rice::String, Rice::Hash>());
  rb_cProducer.define_method("send_message", &Producer::sendMessage);
  rb_cProducer.define_method("send_timer_message", &Producer::sendTimerMessage);
  rb_cProducer.define_method("start", &Producer::start);
  rb_cProducer.define_method("shutdown", &Producer::shutdown);
}
